feat: Pagination for listing data sets and PDS members

traeok opened the pull request:
## Proposed changes

Implements pagination for listing data sets and listing PDS members.

**Behaviors:**
- When expanding a profile node with a search pattern and pagination enabled, navigation controls are added to the top and bottom of the list. The total list of data sets is split into pages and the attributes are only loaded as the user navigates through each page.
- When the user reaches the end of the list, the "Next page" navigation tree item is greyed out and unusable. Conversely, the "Previous page" item is disabled when the user is at the beginning of the list.
- If only one page of items is available, the paginator controls are hidden. Clicking the refresh button or using the context menu option still refreshes the node.
- The full, basic list of data sets / PDS members is only fetched when the user's search pattern changes, when the element (profile node / PDS) is expanded or when the element is refreshed.

This PR also implements a generic Paginator facility in Zowe Explorer API, allowing any Zowe Explorer extender to implement similar functionality for their trees.

### Media (Screenshots & GIFs)

_Note:_ One thing that was added recently and not present in the screenshots is a page count (beside previous page/next page items when enabled).

_(click to expand)_

<details>
<summary><strong>Listing data sets from a profile node, pagination enabled</strong></summary>
<img src="https://i.imgur.com/cSOE4ce.gif" />
</details>

<details>
<summary><strong>Listing PDS members w/ pagination enabled</strong></summary>
<img src="https://i.imgur.com/Kw54GgC.gif" />
</details>

<details>
<summary>It also works when either a PDS or search pattern is favorited 😋</summary>
<img src="https://i.imgur.com/ldIXtYo.png" />
</details>

## Release Notes

Milestone: 3.2.0 or 3.3.0

Changelog:

**ZE:** Added pagination support for data sets and PDS members. Now, when the "Zowe -> Ds: Paginate" setting is enabled, the data sets tree view provides navigation controls and separates the list of items into pages. Users can configure the maximum amount of data sets to list per page by changing the "Zowe -> Ds -> Paginate: Data Sets Per Page" setting.
**ZE API:** Implemented the `Paginator` class, providing a generic facility for extenders to paginate tree nodes within their own tree views.

## Types of changes

<!-- What types of changes does your code introduce to Zowe Explorer? Put an `x` in all boxes that apply. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

<!-- Put an `x` in all boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer. -->

**General**

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] All PR dependencies have been merged and published (if applicable)
- [x] A GIF or screenshot is included in the PR for visual changes
- [x] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`

**Code coverage**

- [x] There is coverage for the code that I have added
- [x] I have added new test cases and they are passing
- [x] I have manually tested the changes

**Deployment**

- [ ] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):

adam-wolfe commented:
This is looking good. I do not see the settings in VS Code, but adding this to settings.json works:
```json
  "zowe.ds.paginate": true,
  "zowe.ds.datasetsPerPage": 5
```

One thing I see, if I go to the next page in the member list of a PDS, I see `1 members with errors`.

traeok commented:
> This is looking good. I do not see the settings in VS Code, but adding this to settings.json works:
> 
> ```json
>   "zowe.ds.paginate": true,
>   "zowe.ds.datasetsPerPage": 5
> ```

Hmm, I may have overwritten their definitions by accident, I'll look into this and follow up.

> One thing I see, if I go to the next page in the member list of a PDS, I see `1 members with errors`.

Thanks @adam-wolfe, I noticed this last night. Talked with Timothy and determined that the `returnedRows` property needed adjusted when the `start` member (cursor for paginator) is filtered from the list. Should be resolved in 74338b0

anaxceron commented on packages/zowe-explorer/CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 - Implemented the `SharedActions.refreshProvider` function to allow refreshing a single tree provider. [#3524](https://github.com/zowe/zowe-explorer-vscode/pull/3524)
 - Added an new VS Code command palette option `Zowe Explorer: Update Profile Configuration Schema`, allowing users to manually trigger an update to global and project level configuration schema files with registered Zowe Explorer extender types. [#1642](https://github.com/zowe/zowe-explorer-vscode/issues/1642)
 - Added script that will fetch latest localizations at every build now. [#3573](https://github.com/zowe/zowe-explorer-vscode/pull/3573)
+- Added pagination support for data sets and PDS members. Now, when the "Zowe -> Ds: Paginate" setting is enabled, the data sets tree view provides navigation controls and separates the list of items into pages. Users can configure the maximum amount of data sets to list per page by changing the "Zowe -> Ds -> Paginate: Data Sets Per Page" setting. [#3585](https://github.com/zowe/zowe-explorer-vs ...
```
```suggestion
- Added pagination support for data sets and PDS members. Now, when the **Zowe -> Ds: Paginate** setting is enabled, the **Data Sets** tree view provides navigation controls and separates the list of items into pages. Users can configure the maximum amount of data sets to list per page by changing the **Zowe -> Ds -> Paginate: Data Sets Per Page** setting. [#3585](https://github.com/zowe/zowe-explorer-vscode/pull/3585)
```

anaxceron reviewed (commented):
Left some suggested edits for one of the changelogs

JillieBeanSim reviewed (changes requested):
thanks for working on this @traeok 
I do see that this doesn't work out the box for extender types, I get the previous and next page but the count doesn't match the setting and all are being displayed.  Since we control the treeviews and data displayed in it and there are no new APIs needing a check for existence like when adding new method to extender or profile type API file it should work for extenders out the box.  Can recreate what I see with the zftp ext on same branch.

traeok commented:
> thanks for working on this @traeok I do see that this doesn't work out the box for extender types, I get the previous and next page but the count doesn't match the setting and all are being displayed. Since we control the treeviews and data displayed in it and there are no new APIs needing a check for existence like when adding new method to extender or profile type API file it should work for extenders out the box. Can recreate what I see with the zftp ext on same branch.

Thanks for the feedback @JillieBeanSim! This is most likely because the extender does not implement the `start` or `maxLength` parameters. For more context, the options for our `allMembers` and `dataSet` functions have an options parameter of type `IListOptions`, which contains both of those parameters. However, if the API does not return data that respects those parameters, then pagination won't work for that profile type.

That said, I'm not sure that our dependency for the FTP extension would support these parameters, but I can add some handling to support those options for z/FTP profiles. Worst case scenario, it just won't be efficient with FTP if FTP doesn't support listing a range of data sets.

adam-wolfe commented:
Two things:
- I'm occasionally seeing an error message: `[Paginator.fetchNextPage] Paginator is already loading.`. I seem to be able to cause this by clicking the `Next page` label repeatedly without waiting for the list to load. It would probably make sense to suppress this message since it has no real impact.
- After changing the `Datasets per page` setting and refreshing the trees, the page count is not updated until the user clicks on the Next/Previous page label. It would be nice if these labels were updated when the tree is refreshed. However, I don't think people would be changing this setting frequently enough for this to really be an issue.

traeok commented:
> Hey @traeok could we have a check of return and if length is higher than setting and can assume it's not supported and just not include the previous/next buttons?

@JillieBeanSim I'm starting to think it would make more sense for there to be a "profile types for pagination" list in the extension settings - we can add z/OSMF and z/FTP by default, and if they have pagination enabled, they can add an extender profile type to the list if they want to use it with that type. If the extender doesn't support those options, they can enhance their API to support them and then the user can add that type to the list for pagination once its ready.

This ensures that the default behavior is kept for other profile types when pagination is enabled and that type isn't added to the list.

traeok commented:
> Two things:
> 
> * I'm occasionally seeing an error message: `[Paginator.fetchNextPage] Paginator is already loading.`. I seem to be able to cause this by clicking the `Next page` label repeatedly without waiting for the list to load. It would probably make sense to suppress this message since it has no real impact.
> * After changing the `Datasets per page` setting and refreshing the trees, the page count is not updated until the user clicks on the Next/Previous page label. It would be nice if these labels were updated when the tree is refreshed. However, I don't think people would be changing this setting frequently enough for this to really be an issue.

Thanks, addressed both in upcoming commit 👍

traeok commented:
> adding more settings makes user experience more difficult, and an opt in setting is strange enough. If this requires adoption from extenders why are we not just making a new API in the extenderAPI file, check for its existence and proceed accordingly?

This is a tough situation and may benefit from further discussion in standup. My suggestion would support extenders out of the box if the options are supported by the APIs (which they should be, since the APIs have been required since v2). The alternative is to make each extender explicitly adopt it, which also duplicates code because the required APIs on the MVS interface should already support those options. I can try to add some logic which overrides the setting if we know that pagination isn't supported, but that limits support out of the box to _only_ z/OSMF and z/FTP.
 
> Also, not sure how we could support zFTP out the box as default if mentioned in previous comment `I'm not sure that our dependency for the FTP extension would support these parameters`

I posted an update on that comment - I added client-side support for these parameters for z/FTP profiles. Even if the underlying API doesn't support it (e.g. zos-node-accessor), the extender's implementation of the function can still do the filtering before it gives us the final response.

JillieBeanSim commented:
> Ok, I'm hoping that my recent changes help to resolve any concern about extender types. If the extender happens to give us a full list, we'll still show that list to the user, and the navigation items are hidden. This keeps the same default behavior of listing all children/data sets, even when pagination is enabled and the extender doesn't support loading ranges.
> 
> **Note:** The following screenshot uses an older version of z/FTP extension that doesn't support `start` or `maxLength`, to show that the default behavior is kept intact for an extender that doesn't support pagination.
> 
> ![image](https://private-user-images.githubusercontent.com/86500639/433418712-02b2faab-3e00-47fe-9187-6fd38631ad58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2NDE5NDIsIm5iZiI6MTc0NDY0MTY0MiwicGF0aCI6Ii84NjUwMDYzOS80MzM0MTg3MTItMDJiMmZhYWItM2UwMC00N2ZlLTkxODctNmZkMzg2MzFhZDU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE0VDE0NDA0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYyZjA3NTBhMTVlOTI3NjY1MWE1MDU5ODYzNTFmYjA1YWM2OWY1MGUyMzljNDI5MWIwYzYwY2M3YmE5ZWQ2YTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZiSpP5mNNlzeCojBcjgDUF9bCiy7537WtmCYVfQAcaQ)

Hey @traeok that looks good, now that we are doing this handling do we really need the opt in setting too or could we just have the paging count setting?  Not having to opt in and seeing this right away when installing latest version would result in a cleaner user experience if we can handle extenders not adopting yet for users.

adam-wolfe commented:
> Hey all, we have numerous bugs (some high priority) that this solution fixes. I don't think we should make users opt into a much needed fix for users with a very large list returned. We can up the default if we think 100 isn't large enough and we can add hover info to the Previous node sharing info to the user on why they are seeing this. Something like `Showing page 1 of ? for search filter ${searchFilter}`

Seems reasonable to me.

zFernand0 reviewed (dismissed):
LGTM! 😋 
---

First, thank you Trae for being so patient in making changes to this PR. I know the discussions dragged this a bit longer than expected. 🙏 

Second, I want to thank everyone who was part of the discussion in a recent ZE standup. I think it was a good way to hash out the default behavior we want. 🙇 

---

Finally, I don't want to throw a wrench in this PR, but I'm seeing something a bit odd. When thinking of sorting dataset members in descending order, I'm afraid that pagination may confuse some folks. It seems that we get the members in ascending order and reverse it in the client. 😕 

For example:
- Assume you have members `A00, A02, ... A99, B00, B01, ... B99, ... Y00, Y01, ... Y99, Z00, Z01, ... Z99`
- With a descending sort order you will see:
  - First page: `A99, A98, ... A00`
  - Second page: `B99, B98, ... B00`
- In reality, we should see:
  - First page: `Z99, Z98, ... Z00`
  - Second page: `Y99, Y98, ... Y00`
  
---

All things considered, I do believe this is a great implementation, which is why I'm approving 🙏 

I'm happy to open a separate issue in case we wanted to look into this further 😋

traeok commented:
> Finally, I don't want to throw a wrench in this PR, but I'm seeing something a bit odd. When thinking of sorting dataset members in descending order, I'm afraid that pagination may confuse some folks. It seems that we get the members in ascending order and reverse it in the client. 😕
> 
> For example:
> 
> * Assume you have members `A00, A02, ... A99, B00, B01, ... B99, ... Y00, Y01, ... Y99, Z00, Z01, ... Z99`
> * With a descending sort order you will see:
>   
>   * First page: `A99, A98, ... A00`
>   * Second page: `B99, B98, ... B00`
> * In reality, we should see:
>   
>   * First page: `Z99, Z98, ... Z00`
>   * Second page: `Y99, Y98, ... Y00`
> 

I think this is to be expected, at least with how sorting is implemented (client-side sorting) and due to z/OSMF limitations @zFernand0 - since it isn't possible to list a range in reverse order from the server. We can only provide the `start` parameter which filters out any data set name that doesn't start with the given text - however, this API assumes that you are traversing in ascending order.

Should we warn users to keep pagination disabled if sorting by descending order using z/OSMF? Alternatively, we can disable it entirely when sort order is set to Name & Descending, but this may be misleading to users. I'm thinking that a warning or note for z/OSMF profiles on Zowe Docs could suffice until z/OSMF supports it (or until we find a way to sort on the server, like with an alternate backend 😋)

JillieBeanSim reviewed (changes requested):
Thanks for working on this @traeok mostly it is looking really good.  I do see an instance when doing filtered search with multiple ds and a member filter like so
bjsimm.jcl*, bjsimm.co*(*copy)
screen shots show what I get 
marketplace (working as expected):
<img width="271" alt="Screenshot 2025-04-17 at 8 55 58 AM" src="https://github.com/user-attachments/assets/a3b37b35-0131-43c2-8182-097992c9dd8a" />
this branch:
<img width="271" alt="Screenshot 2025-04-17 at 8 56 18 AM" src="https://github.com/user-attachments/assets/52d1ab24-e08b-4843-9689-10540cfd70e7" />

traeok commented:
> Thanks for working on this @traeok mostly it is looking really good. I do see an instance when doing filtered search with multiple ds and a member filter like so bjsimm.jcl*, bjsimm.co*(*copy) screen shots show what I get

Thanks Billie, turns out there was a race condition in `datasetFilterPrompt` that removes the filters from the PDSes, resulting in this behavior. I had tested with member patterns a few days ago and it did work, but did limited testing and was able to reproduce once switching between patterns.

Resolved in upcoming commit 👍

zFernand0 commented on packages/zowe-explorer/CHANGELOG.md:
```diff
@@ -4,13 +4,6 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 
 ## TBD Release
 
-### Bug fixes
-
-- BugFix: Fixed an issue where an invalid token in a base configuration profile will take precedence over a plaintext user/password in a service configuration profile. [#3575](https://github.com/zowe/zowe-explorer-vscode/pull/3575)
-- BugFix: Fixed an issue where profiles that were not utilizing tokens were improperly running `checkJwtForProfile()` function within the `ZoweTreeProvider`. [#3575](https://github.com/zowe/zowe-explorer-vscode/pull/3575)
-- Fixed an issue where copy-and-pasting an empty PDS within the same LPAR would result in an error. [#3541](https://github.com/zowe/zowe-explorer-vscode/issues/3541)
-- Added Time Started, Time Ended, and Time Submitted job properties to the Jobs table view. [#3055](https://github.com/zowe/zowe-explorer-vscode/issues/3055)
```
This was just a minor cleanup item.
- https://github.com/zowe/zowe-explorer-vscode/pull/3575#discussion_r2052963701

traeok replied:
Sorry I may be missing something regarding this comment 😅 is there an action I need to take here?

zFernand0 replied:
Nope, nothing here... just clarifying why I did that as part of the branch-merge commit 😋

pujal0909 reviewed (commented):
LGTM! Just wondering, does the settings description need to specify that the number chosen for datasets per page applies to the number of members per page? Or should users have the ability to choose a number for members too?

zFernand0 commented on packages/zowe-explorer/l10n/poeditor.json:
I had some minor changes on this file after running the pre-publish script, but nothing that can't be resolved in a follow-up PR 🙏

zFernand0 commented on packages/zowe-explorer/l10n/bundle.l10n.json:
I had some minor changes on this file after running the pre-publish script, but nothing that can't be resolved in a follow-up PR 🙏

traeok replied:
Resolved in 4b9564a 😋

zFernand0 commented on packages/zowe-explorer-api/__tests__/__unit__/tree/pagination.unit.test.ts:
```diff
@@ -0,0 +1,615 @@
+    const returnLessItemsOnCursor = options?.returnLessItemsOnCursor;
+    const failOnCursorProvided = !!options && Object.prototype.hasOwnProperty.call(options, "failOnCursor");
+
+    return jest.fn(async (cursor: string | undefined, limit: number): Promise<IFetchResult<MockDataItem, string>> => {
```
a few lint warnings in regards to `async` functions without any `await` statements, but that's kind of a minor thing 😅 

Could either disable that lint rule for these or return a `Promise.resolve` 😋

traeok replied:
Sorry about that, addressed in b61fca9

zFernand0 reviewed (approved):
LGTM! 😋 
---

Left some comments but mostly minor stuff. 😅 
Also, it'd be good to run `prettier` just to remove unnecessary empty lines on other files (unrelated to this PR) 😋

traeok commented:
> LGTM! Just wondering, does the settings description need to specify that the number chosen for datasets per page applies to the number of members per page? Or should users have the ability to choose a number for members too?

@pujal0909 For now I think its okay to keep it as one setting, since users can refine their search patterns to narrow down what data sets that they want to interact with. In the interest of avoiding scope creep, we can always add a separate "members per page" setting as a follow-up enhancement. I updated the description in 4b9564a 👍

zFernand0 reviewed (dismissed):
LGTM! 😋 
---

Thanks for addressing some comments already. 😅 

Happy to re-approve if more are coming soon 😋

traeok commented:
> Thanks for addressing some comments already. 😅
> 
> Happy to re-approve if more are coming soon 😋

Sorry Fernando, I keep making you re-approve 😂 should be good to go now.

zFernand0 reviewed (approved):
LGTM! 😋 
---

> Sorry Fernando, I keep making you re-approve 😂 should be good to go now.

nw, 🙏 
Thanks for fixing stuff so quickly 🥳

SanthoshiBoyina1 reviewed (approved):
LGTM. Awesome work and thanks for new feature @traeok !!
